services/wallet/router/router_v2_ut_test.go

```diff
@@ -0,0 +1,145 @@
+	"github.com/status-im/status-go/services/wallet/token"
+	mock_token "github.com/status-im/status-go/services/wallet/token/mock/token"
+)
+
+func setupRouter(t *testing.T) (*Router, mock_token.MockManagerInterface, *MockCandidateResolver, *MockEstimator, *mock_pathprocessor.MockPathProcessor, *mock_pathprocessor.MockPathProcessor, *gomock.Controller) {
+	ctrl := gomock.NewController(t)
+
+	mockTokenManager := mock_token.NewMockManagerInterface(ctrl)
+	mockPathProcessor := mock_pathprocessor.NewMockPathProcessor(ctrl)
+	mockPathProcessor.EXPECT().Name().Return("mockPathProcessor").AnyTimes()
+	mockPathProcessorBridge := mock_pathprocessor.NewMockPathProcessor(ctrl)
+	mockPathProcessorBridge.EXPECT().Name().Return(pathprocessor.ProcessorBridgeHopName).AnyTimes()
+
+	candidateResolver := NewMockCandidateResolver(ctrl)
+	mockEstimator := NewMockEstimator(ctrl)
+	router := NewRouter(nil, nil, mockTokenManager, nil, nil, nil, nil, nil)
+	router.AddPathProcessor(mockPathProcessor)
+	router.AddPathProcessor(mockPathProcessorBridge)
+	router.CandidateResolver = candidateResolver
+	router.Estimator = mockEstimator
+	return router, *mockTokenManager, candidateResolver, mockEstimator, mockPathProcessor, mockPathProcessorBridge, ctrl
+}
+
+func TestRouterResolveCandidatesForNetwork(t *testing.T) {
+	router, mockTokenManager, mockCandidateResolver, _, _, _, ctrl := setupRouter(t)
+	defer ctrl.Finish()
+
+	input := &RouteInputParams{
+		TestnetMode:          false,
+		SendType:             Bridge,
+		AddrFrom:             common.HexToAddress("0x1"),
+		AddrTo:               common.HexToAddress("0x2"),
+		AmountIn:             (*hexutil.Big)(big.NewInt(testAmount1USDC)),
+		TokenID:              pathprocessor.UsdcSymbol,
+		DisabledFromChainIDs: []uint64{},
+		DisabledToChainIDs:   []uint64{},
+	}
+
+	// testParams: &routerTestParams{
+	token := &token.Token{
+		ChainID:  1,
+		Symbol:   pathprocessor.UsdcSymbol,
+		Decimals: 6,
+	}
+
+	networks := []*params.Network{
+		{
+			ChainID: w_common.OptimismMainnet,
+		},
+	}
+
+	network := &params.Network{
+		ChainID: token.ChainID,
+	}
+
+	// Test happy path for bridge hop
+	require.True(t, validateInput(input, network))
+	mockTokenManager.EXPECT().FindToken(network, pathprocessor.UsdcSymbol).Return(token).Times(1)
+	// TODO Define the expected return values - empty for 1st processor, non-empty for 2nd processor
+	mockCandidateResolver.EXPECT().resolveCandidatesForProcessor(gomock.Any(), input, network, token, gomock.Any(), big.NewInt(testAmount1USDC), false, router.pathProcessors[pathprocessor.ProcessorBridgeHopName], networks).Times(1).Return([]*PathV2{
+		{
+			ProcessorName: pathprocessor.ProcessorBridgeHopName,
+		}, // TODO Define the expected return values
+	})
+
+	candidates := router.resolveCandidatesForNetwork(context.TODO(), input, network, networks)
+	require.NotEmpty(t, candidates)
+}
+
+func TestDefaultCandidateResolverResolveCandidateForProcessor(t *testing.T) {
```

Comment:
Testing private function? Aren't they implementation details?

I agree this function is private, but then not expose and not tested., This goes back to the resolveCandidatesForProcessor

Reply:
@igor-sirotin
private method are implementation details, this is not what is being exposed, they should be tested via the public API.
This is all link to the mocking strategy being applied, here, you don't mock private method, actually as I said you mock only external services/package
A valid unit test, is a test i can change the implementation, refactor the code, extract private method, rename things, and the test still pass, if that is not the case and i have to change the test while i didn't change the input/output, the test is not really going to help me in the future

ofc it is best practice, not "rule of thumb" and i understand exception, but clearly here it is more the norm rather than exception.